File: Chris/2022-07-21-Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        prevNodes = {}
        
        
        def hasCycleInner(head):
            if head == None:
                return False

            if head in prevNodes:
                return True
            
            prevNodes[head] = True
            
            return hasCycleInner(head.next)
        
        return hasCycleInner(head)
        

File: HyungJoonLEE/2022-07-21-Linked-List-Cycle.c
#include "common.h"

struct ListNode {
    int val;
    struct ListNode *next;
};

bool hasCycle(struct ListNode *head) {
    struct ListNode *chaser = head, *pioneer = NULL;

    if(!head || !(head->next)) return false;

    pioneer = head->next;

    while(chaser != pioneer) {
        if (pioneer == NULL) return false;

        chaser = chaser->next;
        if(pioneer->next && pioneer->next->next)
            pioneer = pioneer->next->next;
        else
            pioneer = pioneer->next;
    }
    return true;
}
